pkg/workloadmeta/collectors/internal/containerd/image.go

```diff
@@ -30,73 +29,148 @@ import (
+		repoDigestsByID: make(map[string]map[string]struct{}),
 	}
 }
 
-func (images *knownImages) addAssociation(imageName string, imageID string) {
-	images.mut.Lock()
-	defer images.mut.Unlock()
+func (images *knownImages) addReference(imageName string, imageID string) {
+	previousIDReferenced, found := images.idsByName[imageName]
+	if found && previousIDReferenced != imageID {
+		images.deleteReference(imageName, previousIDReferenced)
+	}
 
 	images.idsByName[imageName] = imageID
 
 	if images.namesByID[imageID] == nil {
 		images.namesByID[imageID] = make(map[string]struct{})
 	}
 	images.namesByID[imageID][imageName] = struct{}{}
-}
 
-func (images *knownImages) deleteAssociation(imageName string, imageID string) {
-	images.mut.Lock()
-	defer images.mut.Unlock()
+	if isAnImageID(imageName) {
+		return
+	}
 
+	if isARepoDigest(imageName) {
+		if images.repoDigestsByID[imageID] == nil {
+			images.repoDigestsByID[imageID] = make(map[string]struct{})
+		}
+		images.repoDigestsByID[imageID][imageName] = struct{}{}
+		return
+	}
+
+	// The name is not an image ID or a repo digest, so it has to be a repo tag
+	if images.repoTagsByID[imageID] == nil {
+		images.repoTagsByID[imageID] = make(map[string]struct{})
+	}
+	images.repoTagsByID[imageID][imageName] = struct{}{}
+	return
+}
+
+func (images *knownImages) deleteReference(imageName string, imageID string) {
 	delete(images.idsByName, imageName)
 
-	if images.namesByID[imageID] == nil {
+	if images.namesByID[imageID] != nil {
+		delete(images.namesByID[imageID], imageName)
+	}
+
+	if isAnImageID(imageName) {
 		return
 	}
 
-	delete(images.namesByID[imageID], imageName)
-	if len(images.namesByID[imageID]) == 0 {
-		delete(images.namesByID, imageID)
+	if isARepoDigest(imageName) {
+		if images.repoDigestsByID[imageID] == nil {
+			return
+		}
+		delete(images.repoDigestsByID[imageID], imageName)
+		if len(images.repoDigestsByID[imageID]) == 0 {
+			delete(images.repoDigestsByID, imageID)
+		}
+		return
+	}
+
+	// The name is not an image ID or a repo digest, so it has to be a repo tag
+	if images.repoTagsByID[imageID] == nil {
+		return
+	}
+	delete(images.repoTagsByID[imageID], imageName)
+	if len(images.repoTagsByID[imageID]) == 0 {
+		delete(images.repoTagsByID, imageID)
 	}
 }
 
 func (images *knownImages) getImageID(imageName string) (string, bool) {
-	images.mut.Lock()
-	defer images.mut.Unlock()
-
 	id, found := images.idsByName[imageName]
 	return id, found
 }
 
-func (images *knownImages) isReferenced(imageID string) bool {
-	images.mut.Lock()
-	defer images.mut.Unlock()
+func (images *knownImages) getRepoTags(imageID string) []string {
+	var res []string
+	for repoTag := range images.repoTagsByID[imageID] {
+		res = append(res, repoTag)
+	}
+	return res
+}
+
+func (images *knownImages) getRepoDigests(imageID string) []string {
+	var res []string
+	for repoDigest := range images.repoDigestsByID[imageID] {
+		res = append(res, repoDigest)
+	}
+	return res
+}
```

Comment:
Do we really need to copy the slices ?
Can’t we just return them like:

```suggestion
func (images *knownImages) getRepoTags(imageID string) []string {
	return images.repoTagsByID[imageID]
}

func (images *knownImages) getRepoDigests(imageID string) []string {
	return images.repoDigestsByID[imageID]
}
```

If not, we can use `copy(…)`:

```suggestion
func (images *knownImages) getRepoTags(imageID string) []string {
	res := make([]string, len(images.repoTagsByID[imageID]))
	copy(res, images.repoTagsByID[imageID])
	return res
}

func (images *knownImages) getRepoDigests(imageID string) []string {
	res := make([]string, len(images.repoDigestsByID[imageID]))
	copy(res, images.repoDigestsByID[imageID])
	return res
}
```

Reply:
That's not going to work in this case because `images.repoTagsByID` is a set `map[string]struct{}` (to handle duplicates) and the function needs to return `[]string` (which is what `workloadmeta.ContainerImageMetadata` accepts .